File: 0822/경주로 건설.js
var minPrice;
var n;
const dir = {
    'U': [-1, 0],
    'R': [0, 1],
    'D': [1, 0],
    'L': [0, -1]
}

solution([[0,0,0],[0,0,0],[0,0,0]]);
function solution(board) {
    var answer = 0;
    n = board.length;
    minPrice = n * n * 500;

    let map = board;
    map[0][0] = 1;
    map[0][1] = 1;
    dfs(map, [0, 1], 100, 'R');
    
    map = board;
    map[0][0] = 1;
    map[1][0] = 1;
    dfs(map, [1, 0], 100, 'D');
    
    answer = minPrice;
    return answer;
}

function dfs(sMap, cur, price, prevDir){
    console.log(cur)
    if (cur[0] === n-1 && cur[1] === n-1){
        console.log(price)
        minPrice = Math.min(price, minPrice);
    }
    const sRow = cur[0];
    const sCol = cur[1];
    let newPrice;
    let newMap;
    sMap[sRow][sCol] = 1;
    console.log(sMap)
    //L
    if (sCol > 0 && sMap[sRow][sCol-1] === 0){
        newPrice = prevDir === 'L' ? 100 : 600;
        newMap = sMap;
        newMap[sRow][sCol-1] = 1;
        dfs(newMap, [sRow, sCol-1], price + newPrice, 'L')
    }
    //D
    if (sRow < n - 1 && sMap[sRow + 1][sCol] === 0){
        newPrice = prevDir === 'D' ? 100 : 600;
        newMap = sMap;
        newMap[sRow + 1][sCol] = 1;
        dfs(newMap, [sRow+1, sCol], price + newPrice, 'D')
    }
    //R
    if (sCol < n - 1 && sMap[sRow][sCol+1] === 0){
        newPrice = prevDir === 'R' ? 100 : 600;
        newMap = sMap;
        newMap[sRow][sCol+1] = 1;
        dfs(newMap, [sRow, sCol+1], price + newPrice, 'R')
    }
    //U
    if (sRow > 0 && sMap[sRow-1][sCol] === 0){
        newPrice = prevDir === 'U' ? 100 : 600;
        newMap = sMap;
        newMap[sRow-1][sCol] = 1;
        dfs(newMap, [sRow-1, sCol], price + newPrice, 'U')
    }
}

File: 0822/경주로 건설.java
import java.util.*;
class Solution {

    private static int N;
    private static boolean[][] visited;
    private static int[][] map;

    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, 1, 0, -1};

    private static int cost = Integer.MAX_VALUE;

    public int solution(int[][] board) {
        int answer = 0;
        N = board.length;

        visited = new boolean[N][N];
        map = board;

        // System.out.println(cost);

        BFS(0,0,-1,0);

        answer = cost;

        return answer;
    }


    private static void BFS(int x, int y, int dir, int price) {
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(x, y, dir, price));
        visited[x][y] = true;

        while(!q.isEmpty()) {
            Node node = q.remove();

            int cx = node.x;
            int cy = node.y;
            int cDir = node.dir;
            int cPrice = node.price;

            // 도착
            if (cx == N-1 && cy == N-1) {
                cost = Math.min(cost, cPrice);
            }

            for (int i = 0 ; i < dx.length; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                int nDir = i;
                int nPrice = cPrice;

                // 이동 불가능한 경우
                if (nx < 0 || nx >= N || ny < 0 || ny >= N || map[nx][ny] == 1) {
                    continue;
                }

                // 첫 이동이거나 같은 방향일 때
                if (cDir == -1 || cDir == nDir) {
                    nPrice += 100;
                } else {
                    // 방향이 다를 때
                    nPrice += 600;
                }

                if (!visited[nx][ny] || map[nx][ny] >= nPrice) {
                    visited[nx][ny] = true;
                    map[nx][ny] = nPrice;
                    q.add(new Node(nx, ny, nDir, nPrice));
                }
            }
        }
    }
}

class Node{
    int x;
    int y;
    int price;
    int dir;

    public Node(int x, int y, int dir, int price) {
        this.x = x;
        this.y = y;
        this.price = price;
        this.dir = dir;
    }
}
>>>>>>> 3fb6e50a6a8e0fe7eb3490f6d730e83ef327c371
